Adds filesize component

robmarkcole opened the pull request:
## Description:
Component for displaying the size of a file.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io/pull/4589) with documentation (if applicable):** home-assistant/home-assistant.github.io#4589

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: filesize
    file_paths:
      - /config/home-assistant_v2.db
```

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x ] Tests have been added to verify that the new code works.

robmarkcole committed:
Create filesize.py

arsaboo commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,92 @@
+        """Get the size of the file."""
+        self._size = self.get_file_size(self._path)
+
+    def get_file_size(self, path):
```
Add a docstring here.

robmarkcole replied:
Added

arsaboo commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,92 @@
+        file_size = round(statinfo.st_size/1e6, decimals)
+        return file_size
+
+    def get_last_updated(self, path):
```
Same...missing docstring.

robmarkcole replied:
Added

robmarkcole committed:
Update filesize.py

fabaff commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+        self._size = self.get_file_size(self._path)
+
+    def get_file_size(self, path):
+        """Returns the size of the file in MB."""
```
Please use `Return` instead of `Returns`. PEP257 checks flag this as violation.

robmarkcole replied:
ok

fabaff commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+        """Return other details about the sensor state."""
+        attrs = {}
+        attrs['path'] = self._path
+        attrs['last_updated'] = self.get_last_updated(self._path)
```
Call `get_last_updated()` in `update()`.

robmarkcole replied:
ok

fabaff reviewed (commented):
It would also require a test.

fabaff commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+        """Returns the time the file was last modified."""
+        statinfo = os.stat(path)
+        last_updated = datetime.datetime.fromtimestamp(statinfo.st_mtime)
+        last_updated = last_updated.strftime(DATE_STR_FORMAT)
```
Return an ISO 8601 timestamp.

robmarkcole replied:
OK gone with ```last_updated = last_updated.isoformat(' ')``` as the default adds an ugly 'T'

Qxlkdr commented:
Don't forget to add your file to .coveragerc as described in the pull request template.

robmarkcole commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+import datetime
+import logging
+import os
+import voluptuous as vol
```
ok

robmarkcole commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+CONF_FILE_PATHS = 'file_paths'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_FILE_PATHS): [cv.isfile],
```
ok

robmarkcole commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+
+
+class Filesize(Entity):
+    """Representation of the HA database."""
```
OK

robmarkcole commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+        self._last_updated = None
+        self._name = path.split("/")[-1]
+        self._unit_of_measurement = 'MB'
+        self.update()
```
ok

robmarkcole committed:
Updates filesize

Addresses issues raised in review

robmarkcole commented:
@Qxlkdr I don't appear to be able to edit more than this one file since I created this pull via the 'create file' button. Will I need to create a branch, add this file and edit coverage in the branch, then create a new PR?

![image](https://user-images.githubusercontent.com/11855322/35902728-39e9cd22-0bd4-11e8-8b32-0febae9ff05b.png)

@fabaff agreed a test is required but request advice on how to test - use a file in this repo that will not change over time?

arsaboo commented:
@robmarkcole You can modify the files in your `patch-1` fork to keep adding to this PR. https://github.com/robmarkcole/home-assistant/tree/patch-1.

robmarkcole committed:
Update .coveragerc

fabaff commented:
The `file` sensor has a test. I guess that it would be possible to use that as base.

robmarkcole commented:
@fabaff yes looks like I just need to mock the return of  os.stat() [similiar ](https://github.com/home-assistant/home-assistant/blob/acb521330c7aa27dacc62851f96649cfdbb377bc/tests/components/sensor/test_file.py#L74)to the file test

balloob commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+    """Set up the file size sensor."""
+    sensors = []
+    for path in config.get(CONF_FILE_PATHS):
+        sensors.append(Filesize(path))
```
All paths have to be whitelisted by checking it with `hass.config.is_allowed_path(path)`

robmarkcole replied:
Done

MartinHjelmare commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,94 @@
+    @property
+    def device_state_attributes(self):
+        """Return other details about the sensor state."""
+        attrs = {}
```
Please construct the whole dict directly.

robmarkcole committed:
Addresses requested changes

Addresses the changes requested by @baloob.
Additionally the file size in bytes is now available in attributes.

robmarkcole committed:
Create test_filesize.py

This isn't working yet

robmarkcole commented:
I could use some advice on the test. I'm using a path to a file on my machine, then getting error:
```
    @patch('os.stat', Mock(return_value=FAKE))
    def test_filesize_class(self):
        self.assertTrue(
            setup_component(self.hass, 'sensor', {'sensor': self.config}))

        state = self.hass.states.get('sensor.homeassistant_v2db')

>       assert state.state == 37.99
E       AttributeError: 'NoneType' object has no attribute 'state'
```

Also probably theres a more appropriate way than adding the class FakeObj() ?
Cheers

robmarkcole committed:
Update test_filesize.py

robmarkcole commented:
I've had another play but still unsure why this test is failing. Appears the sensor is created successfully but then the nonetype error. Any suggestions? @arsaboo @fabaff  @dgomes

robmarkcole committed:
Update test_filesize.py

robmarkcole committed:
Update test_filesize.py

robmarkcole committed:
Update test_filesize.py

robmarkcole committed:
Update test_filesize.py

robmarkcole committed:
Update test_filesize.py

dgomes committed:
fixed valid file test

robmarkcole committed:
Merge pull request #2 from dgomes/filesize

fixed valid file test

robmarkcole committed:
Update test_filesize.py

robmarkcole committed:
Fix indentation

Fix incorrect indentation in setup

robmarkcole commented:
Travis is still complaining about my log error message string formatting - assume we should be using the new {} style?

robmarkcole committed:
Update filesize.py

robmarkcole commented:
@MartinHjelmare @arsaboo I've made all the required changes now, don't know what that strange error from Travis is about. Cheers

balloob commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,105 @@
+        if not hass.config.is_allowed_path(path):
+            _LOGGER.error(
+                "Filepath %s is not valid or allowed", path)
+            return
```
Change to `continue` so correct sensors will still be setup.

robmarkcole replied:
ok

balloob commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,105 @@
+    for path in config.get(CONF_FILE_PATHS):
+        if not hass.config.is_allowed_path(path):
+            _LOGGER.error(
+                "Filepath %s is not valid or allowed", path)
```
Make sure to make a note in the docs how to change allowed paths

robmarkcole replied:
Done. For whitelist_external_dirs I've linked https://home-assistant.io/docs/configuration/basic/ but perhaps theres a better page to point to?

balloob commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,105 @@
+
+def get_last_updated(path):
+    """Return the time the file was last modified."""
+    statinfo = os.stat(path)
```
You're calling `os.stat` twice per update because you made 2 functions. Why not just keep this all inside the `update` method?

robmarkcole replied:
Good point, done

balloob commented on homeassistant/components/sensor/filesize.py:
```diff
@@ -0,0 +1,105 @@
+    """Return the time the file was last modified."""
+    statinfo = os.stat(path)
+    last_updated = datetime.datetime.fromtimestamp(statinfo.st_mtime)
+    last_updated = last_updated.isoformat(' ')
```
Use `isoformat()`

robmarkcole replied:
ok

balloob commented on tests/components/sensor/test_filesize.py:
```diff
@@ -0,0 +1,58 @@
+TEST_FILE = os.path.join(TEST_DIR, 'mock_file_test_filesize.txt')
+
+
+def create_file(path):
```
Use the `tempfile` module that is part of Python.

balloob replied:
Actually, it's fine.

robmarkcole replied:
ok

robmarkcole committed:
Update filesize.py
